api/xds_protocol.rst

```diff
@@ -469,3 +469,155 @@ occurred via a :ref:`resource update <xds_protocol_resource_update>`.
+uses the delta xDS protocol to allow a route configuration to be subscribed to
+and the necessary virtual hosts to be requested as needed. Instead of sending
+all virtual hosts with a route config, using VHDS will allow an Envoy instance
+to subscribe and unsubscribe from a list of virtual hosts stored internally in
+the xDS management server. The xDS management server will monitor this list and
+use it to filter the configuration sent to an individual Envoy instance to only
+contain the subscribed virtual hosts.
+
+In VHDS, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with the 
+:ref:`type_url <envoy_api_field_DeltaDiscoveryRequest.type_url>` set to 
+`type.googleapis.com/envoy.api.v2.route.VirtualHost` 
+and :ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>` 
+set to a list of virtual host resource names for which it would like
+configuration. The management server will respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+with the :ref:`resources <envoy_api_field_DeltaDiscoveryResponse.resources>`
+field populated with the specified virtual hosts, the 
+:ref:`name <envoy_api_field_resource.name>` field populated with
+the virtual host name, and the 
+:ref:`aliases <envoy_api_field_resource.aliases>` field
+populated with the explicit (no wildcard) domains of the virtual host. Future
+updates to these virtual hosts will be sent via spontaneous updates.
+
+Updates to the route configuration entry to which a virtual host belongs will
+clear the virtual host table and require all virtual hosts to be sent again. It
+may be useful for the management server to populate RDS responses with the
+subscribed list of virtual hosts. 
+
+Joining/Reconnecting to a management server
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the current
+namespace. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Virtual Host Resource Naming Convention
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Virtual hosts in VHDS are identified by a combination of the name of the route
+configuration to which the virtual host belong as well as the host HTTP "host"
+header (authority for HTTP2) entry. Resources should be named as follows:
+
+<route configuration name>/<host entry>
+
+Note that matching should be done from right to left since a host entry cannot
+contain slashes while a route configuration name can. 
+
+Requesting Additional resources
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Envoy will send a
```

Comment:
This paragraph seems to duplicate most of the one starting on line 494 above.

And actually, covering the specific details of the incremental protocol seems redundant anyway, since that's already covered earlier in the document.  Maybe all we need to do here is to say that the VHDS uses the incremental protocol and describe how the resource names and resource aliases work for VHDS.

Reply:
I believe it was a request from earlier on that implementation details were covered in the docs.

Reply:
I think this is resolved now.

Reply:
Rewinding back to the beginning of the conversation. 

I'm not sure where this came from:

> Okay, so it sounds like you are actually expecting the client to magically have an implicit VHDS subscription for any virtual host present in the RDS response, even though the client never sent a VHDS subscription message for those resources.

It *is* possible to simultaneously have vhosts communicated via RDS and VHDS, just not the same vhosts, the main differences are that of granularity (route_config vs vhost) and support for on-demand updates (VHDS only). Vhosts in RDS and VHDS are quite separate from each other and both sets are merged when the route matcher is constructed.

I assume this:

> We could have the RDS response include a list of virtual hosts for the client to automatically subscribe to via VHDS, and the client would send a VHDS subscription request as soon as it got the RDS response. This would cause one more round-trip before the data gets to the client on startup, but it would avoid sending the data twice.

came from the "implict VHDS subscriptions" bit, I don't think it makes sense considering how RDS and VHDS handle vhosts. Additionally, I should mention that we can't discard the overhead of sending vhosts twice, one of the main goals VHDS was to minimise the overhead, and this implementation (out of several) was picked specifically b/c it has the lowest overhead.

Reply:
I think it would require minimal changes to verify if vhds.config_source has changed in an rds update, and skip vhds subscription creation if it hasn't. That would eliminate re-sending of VHDS state on a subscription initialization/startup. As this is a protocol change, should do it now.

@htuch -- thoughts?

Reply:
It looks like some of this discussion somehow wound up here instead of in the thread above.  But see my response there (https://github.com/envoyproxy/envoy/pull/7415#discussion_r353847912) for my thoughts.